README.md
# cisco_acl
Python module for working with Cisco ACLs


HISTORY.rst
=======
History
=======

0.1.0 (2017-06-05)
------------------

* First release on github.


tests/test_acl_audit.py
import os.path
from cisco_acl.acl_audit import AclAuditor


def test_acl_audit():
    aclfile = os.path.join(os.path.dirname(__file__), 'data/acl2')
    a = AclAuditor(acl=aclfile)
    assert a.errors[9].startswith('Invalid ACE')
    assert a.errors[11].startswith('Invalid subnet')
    assert a.errors[12].startswith('Invalid host IP')
    assert a.errors[13].startswith('Invalid port')
    assert len(a.errors) == 4


tests/test_acl_regexes.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_acl_regexes
----------------------------------

Tests for `regexes` module.
"""

import os
from cisco_acl.regexes import ace_match


def test_acl_regexes():
    acl_file = os.path.join(os.path.dirname(__file__), 'data/acl1')

    with open(acl_file) as f:
        acl_lines = f.readlines()

    for line in acl_lines:
        # allow for comments
        if line.startswith('#') or line.startswith('!'):
            continue
        assert ace_match(line)


def test_bad_ace():
    ace = 'permit tcp any host eq 80'
    assert ace_match(ace) is False


tests/test_acl_convert_mask.py
from cisco_acl.convert_mask import translate_mask


def test_convert_mask():
    aces_with_wildcard = [
        'permit tcp 172.30.1.0 0.0.0.255 any eq 443',
    ]

    aces_with_hostmask = [
        'permit tcp 172.30.1.0 255.255.255.0 any eq 443',
    ]

    aces_with_cidr = [
        'permit tcp 172.30.1.0/24 any eq 443'
    ]

    assert translate_mask(aces_with_wildcard, 'wc', 'cidr')[0] == aces_with_cidr[0]
    assert translate_mask(aces_with_wildcard, 'wc', 'subnet')[0] == aces_with_hostmask[0]
    assert translate_mask(aces_with_hostmask, 'subnet', 'cidr')[0] == aces_with_cidr[0]
    assert translate_mask(aces_with_hostmask, 'subnet', 'wc')[0] == aces_with_wildcard[0]


cisco_acl/regexes.py
"""
Cisco ACL regex helper library for parsing IOS/ASA ACLs

Examples:
    >>> re.match(ip_address_rx, '1.1.1.1')
    <_sre.SRE_Match object; span=(0, 7), match='1.1.1.1'>
    >>> re.match(ip_address_rx, '2001:420:210d::a')
    <_sre.SRE_Match object; span=(0, 16), match='2001:420:210d::a'>
    >>> ace_match('permit tcp any host 1.1.1.1 eq 80')['destination']
    'host 1.1.1.1'
    >>> ace_match('permit tcp any host 2001:420:210d::a eq 80')['destination']
    'host 2001:420:210d::a'
    >>> ace_match('permit tcp any host www.cisco.com eq 80')['destination']
    'host www.cisco.com'
    >>> re.match(host_or_network_rx, 'host 1.1.1.1')
    <_sre.SRE_Match object; span=(0, 12), match='host 1.1.1.1'>
    >>> re.match(host_or_network_rx, 'host 2001:420:210d::a')
    <_sre.SRE_Match object; span=(0, 21), match='host 2001:420:210d::a'>
"""
import re

# Regular expressions for parsing IPv4/6 addresses
ipv4_address_rx = (
    r'(?:'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])'
)
ipv6_address_rx = (
    r'(?:'
    r'(?:[0-9A-Fa-f]{1,4}\:){6}'
    r'(?:[0-9A-Fa-f]{1,4}\:[0-9A-Fa-f]{1,4}|'
    r'(?:'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))|\:\:'
    r'(?:[0-9A-Fa-f]{1,4}\:){5}(?:[0-9A-Fa-f]{1,4}\:[0-9A-Fa-f]{1,4}|'
    r'(?:'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))|'
    r'(?:[0-9A-Fa-f]{1,4})?\:\:(?:[0-9A-Fa-f]{1,4}\:){4}'
    r'(?:[0-9A-Fa-f]{1,4}\:[0-9A-Fa-f]{1,4}|'
    r'(?:'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))|'
    r'(?:[0-9A-Fa-f]{1,4}\:[0-9A-Fa-f]{1,4})?\:\:'
    r'(?:[0-9A-Fa-f]{1,4}\:){3}'
    r'(?:[0-9A-Fa-f]{1,4}\:[0-9A-Fa-f]{1,4}|'
    r'(?:'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))|'
    r'(?:'
    r'(?:[0-9A-Fa-f]{1,4}\:){,2}[0-9A-Fa-f]{1,4})?\:\:'
    r'(?:[0-9A-Fa-f]{1,4}\:){2}(?:[0-9A-Fa-f]{1,4}\:[0-9A-Fa-f]{1,4}|'
    r'(?:'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))|'
    r'(?:'
    r'(?:[0-9A-Fa-f]{1,4}\:){,3}[0-9A-Fa-f]{1,4})?\:\:[0-9A-Fa-f]{1,4}\:'
    r'(?:[0-9A-Fa-f]{1,4}\:[0-9A-Fa-f]{1,4}|'
    r'(?:'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))|'
    r'(?:'
    r'(?:[0-9A-Fa-f]{1,4}\:){,4}[0-9A-Fa-f]{1,4})?\:\:'
    r'(?:[0-9A-Fa-f]{1,4}\:[0-9A-Fa-f]{1,4}|'
    r'(?:'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}'
    r'(?:[0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))|'
    r'(?:'
    r'(?:[0-9A-Fa-f]{1,4}\:){,5}[0-9A-Fa-f]{1,4})?\:\:[0-9A-Fa-f]{1,4}|'
    r'(?:'
    r'(?:[0-9A-Fa-f]{1,4}\:){,6}[0-9A-Fa-f]{1,4})?\:\:)'
)

ip_address_rx = r'({0}|{1})'.format(ipv4_address_rx, ipv6_address_rx)
subnet_rx = r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\s+\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}'

# Regular expressions for processing Cisco ACL data
# DNS names using the 'host' keyword
dnsname_rx = r'[A-Za-z0-9\-\.\_]+'
# Remarks in Cisco ASA
asa_remark = r'(^access-list\s+([A-Za-z0-9\-_]+)\s+(?:\s+line\s+\d+\s+)?extended\s+remark)'

# ACE protocol
protocol_rx = (
    r'(?:ahp|eigrp|esp|gre|icmp|igmp|ip|object-group|ospf|pcp|pim|tcp|udp|\d{1,3})'
    r'|'
    r'(?:object-group\s+[A-Za-z0-9\-\.\_\:]+)'
)

# ACE source/destination network
host_or_network_rx = (
    r'(?:host\s+({ip_rx}|[A-Za-z0-9\-\.]+))'
    r'|'
    r'(?:({net_rx}))'
    r'|'
    r'(?:any|any4)'
    r'|'
    r'(?:object-group\s+[A-Za-z0-9\-\.\_\:]+)'
    r'|'
    r'(?:addrgroup\s+[A-Za-z0-9\-\.\_]+)'
    r'|'.format(ip_rx=ip_address_rx, net_rx=subnet_rx)
)

# ACE source/destination ports
ports_rx = (
    r'(?:(?:e(?:q)?|gt|lt|ne|le|ge)\s+(?:\d{1,5}|[A-Za-z0-9\-](?:\s+)?)+)'
    r'|'
    r'(?:r(a|an|ang|ange)?\s+(?:[A-Za-z0-9\-]+)\s+(?:[A-Za-z0-9\-]+))'
    r'|'
    r'(?:object-group\s+[A-Za-z0-9\-\.\_]+)'
    r'|'
    r'(?:port-group\s+[A-Za-z0-9\-\.\_]+)'
)

# ACE keywords
keyword_rx = (
    r'(?:established|echo(?:-reply)?|time-exceeded|unreachable|log)'
)

# Full ACE regex
cisco_acl_regex = (
    r'((?P<sequence>\d+)\s+)?'
    r'(?:access-list\s+(?P<name>[A-Za-z0-9\-\_]+)\s+(?:\s+line\s+\d+\s+)?extended\s+)?'
    r'(?P<action>permit|deny)'  # action
    r'\s+'
    r'(?P<protocol>{pro_rx})'  # protocol
    r'\s+'
    r'(?P<source>{net_rx})'  # source network
    r'(?:\s+(?P<source_ports>{prt_rx}))?'  # source ports
    r'\s+'
    r'(?P<destination>{net_rx})'  # destination network
    r'(?:\s+)?'
    r'(?:\s+(?P<destination_ports>{prt_rx}))?'  # destination ports
    r'(?:\s+(?P<keyword>{key_rx}))?'  # keywords (ex. established)
    r'(?:\(hitcnt=\d+\)\s+[A-Za-z0-9]+)?'  # (hitcnt=0) 0x0540b3cb
    r'$'.format(
        pro_rx=protocol_rx,
        net_rx=host_or_network_rx,
        prt_rx=ports_rx,
        key_rx=keyword_rx
    )
)


# Simple function for checking if an ACE matches our regexes above
def ace_match(ace):
    """
    Check if an ACE matches our ACE regex

    Args:
        ace (str): Access control entry

    Returns:
        permission (dict): dictionary of permission details, false otherwise

    Examples:
        >>> from pprint import pprint
        >>> pprint(ace_match('permit tcp any any eq 443'))
        {'action': 'permit',
         'destination': 'any',
         'destination_ports': 'eq 443',
         'keyword': None,
         'name': None,
         'protocol': 'tcp',
         'sequence': None,
         'source': 'any',
         'source_ports': None}
    """
    re_obj = re.compile(cisco_acl_regex, re.I)
    match = re_obj.match(ace.lower())
    if match:
        return match.groupdict()
    else:
        return False


cisco_acl/acl_audit.py
"""
Audit an ACL, checking for the following:
* Invalid ACEs (syntax errors)
* Networking errors (invalid IP/subnets)
* Invalid ports

Supported ACL formats:
* IOS extended
* ASA extended
"""
import logging
import os.path
import re
from ipaddress import ip_network, ip_address
from cisco_acl.regexes import ace_match, ip_address_rx, subnet_rx, dnsname_rx, keyword_rx
from cisco_acl.port_translations import translate_port

logging.getLogger(__name__)


class AclAuditor:
    def __init__(self, **kwargs):
        self.acl = kwargs.get('acl')
        self.acl_format = kwargs.get('format', 'ios')
        self.supported_formats = [
            'ios-extended',
            'asa-extended'
        ]
        self.aces = {}
        """
        aces = {
            # line_num: ace
            1: 'permit tcp any any eq 80'
        }
        """
        self.permissions = {}
        """
        permissions = {
            # line_num: perm
            1: {
            'action': 'permit',
            'destination': 'any',
            'destination_ports': 'eq 80',
            'keyword': None,
            'name': None,
            'protocol': 'tcp',
            'sequence': None,
            'source': 'any',
            'source_ports': None},
        }
        """
        self.errors = {}
        """
        errors = {
            # line_num: error
            5: 'Invalid subnet 1.1.5.5/28: Host bits set',
            10: 'Invalid ACE syntax',
        }
        """
        self._parse()
        self._run_audit()

    def _parse(self):
        acl = os.path.abspath(self.acl)
        if not os.path.isfile(acl):
            raise FileNotFoundError

        with open(acl, mode='rt', errors='ignore', encoding='utf-8') as f:
            for i, line in enumerate(f.readlines(), start=1):
                line = line.strip()
                if line == '':  # Skip blank lines
                    continue

                if line.startswith('!'):  # Skip comments
                    continue

                if line.startswith('remark'):
                    self.aces[i] = line
                    continue
                ace = ace_match(line)
                if not ace:
                    self.errors[i] = 'Invalid ACE: ' + line
                else:
                    self.permissions[i] = ace
                    self.aces[i] = line

    def _run_audit(self):
        logging.info('Processing networking errors ...')
        for i, perm in self.permissions.items():
            self._audit_networks({i: perm})
            self._audit_ports({i: perm})

    def _audit_networks(self, permission):
        for i, perm in permission.items():
            for net in [perm['source'], perm['destination']]:
                if net == 'any':
                    continue

                if net.startswith('object-group') or net.startswith('addrgroup'):
                    og = net.split()[1]
                    if not re.match(dnsname_rx, og):
                        self.errors[i] = 'Invalid object-group: ' + og
                    else:
                        continue

                if net.startswith('host'):
                    host = net.split()[1]

                    if re.match(ip_address_rx, host):
                        try:
                            ip = ip_address(host)
                        except ValueError:
                            self.errors[i] = 'Invalid host IP: ' + host
                    else:
                        if re.match('\d{1,3}.\d{1,3}.\d{1,3}.\d{1,3}', host):
                            self.errors[i] = 'Invalid host IP: ' + host
                        elif not re.match(dnsname_rx, host):
                            self.errors[i] = 'Invalid host: ' + host

                elif re.match(subnet_rx, net):
                    try:
                        network = ip_network('/'.join(net.split()))
                    except ValueError as e:
                        self.errors[i] = 'Invalid subnet "{0}": {1}'.format(net, e)

                else:
                    self.errors[i] = 'Invalid host/network: {0}'.format(net)

    def _audit_ports(self, permission):
        for i, perm in permission.items():
            if perm['protocol'].lower() not in ['tcp', 'udp']:
                continue
            for ports in [perm['source_ports'], perm['destination_ports']]:
                if ports is None:
                    continue
                for p in ports.split()[1:]:
                    if not re.match('\d+$', p):
                        if re.match(keyword_rx, p):
                            continue
                        port_num = translate_port('ios', perm['protocol'], [p], 'to_number')[0]
                        if p == port_num:
                            self.errors[i] = 'Invalid port: {0} - {1} {2}'.format(self.acl_format, perm['protocol'], p)


cisco_acl/port_translations.py
"""
  This library is meant to assist with translating port numbers to port names (and vice versa)
in Cisco IOS and ASA formatted access-lists

Port types that require translations
* eq           Match only packets on a given port number
* gt           Match only packets with a greater port number
* ge           Match only packets with a greater or equal port number
* lt           Match only packets with a lower port number
* le           Match only packets with an equal or lower port number
* neq          Match only packets not on a given port number
* range        Match only packets in the range of port numbers

Examples:

# Import the library:
# from cisco_acl.port_translations import PortTranslator

# Translate an IOS ACE from port numbers to port names
>>> ace = 'permit tcp any any eq 80'
>>> PortTranslator(ace).translate_ace(acl_format='ios', conversion_type='to_name')
'permit tcp any any eq www'

# Translate an IOS ACE from port names to port numbers
>>> ace = 'permit tcp any any range ftp-data ftp'
>>> PortTranslator(ace).translate_ace(acl_format='ios', conversion_type='to_number')
'permit tcp any any range 20 21'

# Translate an ASA ACE from port names to numbers
>>> ace = 'access-list test extended permit tcp any host 64.104.140.141 eq ssh'
>>> PortTranslator(ace).translate_ace(acl_format='asa', conversion_type='to_number')
'access-list test extended permit tcp any host 64.104.140.141 eq 22'

# Translate ASA ACE from port numbers to names
>>> ace = 'access-list test extended permit tcp any any range 80 81'
>>> PortTranslator(ace).translate_ace(acl_format='asa', conversion_type='to_name')
'access-list test extended permit tcp any any range www 81'
"""
import re
import json
import os
import logging
from cisco_acl.regexes import ace_match

logging.getLogger(__name__)

translation_file = os.path.join(os.path.dirname(__file__), 'port_translations.json')
with open(translation_file, mode='rt') as f:
    translation_groups = json.loads(f.read())


def translate_port(acl_format, protocol, ports, conversion_type):
    """
    Translate a port from name to number or vice versa 

    Args:
        acl_format (str): 'ios' or 'asa' 
        protocol (str): protocol from ACE
        ports (str): ports from ace (ex. '80', '80 443')
        conversion_type (str): 'to_name' or 'to_number'

    Returns:
        list: translated ports
    """

    def translate(port):
        if conversion_type == 'to_name':
            for port_name, num in translation_groups[acl_format][protocol].items():
                try:
                    if int(num) == int(port):
                        return port_name
                except:
                    return port

        elif conversion_type == 'to_number':
            logging.debug(list(translation_groups[acl_format][protocol].keys()))
            if port in translation_groups[acl_format][protocol]:
                return translation_groups[acl_format][protocol][port]

        return port

    translated_ports = [translate(port) for port in ports]

    if not translated_ports:
        return ports
    else:
        return translated_ports


class PortTranslator:
    def __init__(self, ace):
        self.ace = ace
        self.formats = ['ios', 'asa']
        self.permission = dict()
        self._parse_ace()

    def _parse_ace(self):
        """
        Parse an ACE using the cisco_acl.regexes library
        """
        permission = ace_match(self.ace)
        if not permission:
            raise SyntaxError('Invalid ACE: {0}'.format(self.ace))
        self.permission = permission

    def translate_ace(self, acl_format='ios', conversion_type='to_name'):
        """
        Translate ports in an ACE between names and numbers
        
        Args:
            acl_format (str): 'ios' or 'asa' 
            conversion_type (str): 'to_name' or 'to_number' 

        Returns:
            str: ace with ports translated
        """
        if self.permission['protocol'].lower() not in ['tcp', 'udp']:
            return self.ace

        if acl_format not in translation_groups:
            raise ValueError('ACL format "{0}" not in {1}'.format(acl_format, list(translation_groups.keys())))

        conversion_types = ['to_name', 'to_number']
        if conversion_type not in conversion_types:
            raise ValueError('Unknown conversion type: {0} Acceptable types: {1}'.format(
                conversion_type, conversion_types))

        # We have two possible conversions to make (source ports, destination ports)
        ports = [self.permission['source_ports'], self.permission['destination_ports']]

        line = self.ace

        for port in ports:
            if port is None:
                continue

            if re.match('^object-group', port, flags=re.I):
                continue

            port = port.lower()
            _ports = port.split()[1:]
            _translated_ports = translate_port(
                acl_format, self.permission['protocol'].lower(), _ports, conversion_type
            )
            if len(_translated_ports) > 1:
                _translated_ports = ' '.join(_translated_ports)
            else:
                _translated_ports = _translated_ports[0]

            if port.startswith('e'):  # eq
                line = re.sub(port, 'eq {}'.format(_translated_ports), line, flags=re.I)
            elif port.startswith('gt'):  # gt
                line = re.sub(port, 'gt {}'.format(_translated_ports), line, flags=re.I)
            elif port.startswith('ge'):  # ge
                line = re.sub(port, 'ge {}'.format(_translated_ports), line, flags=re.I)
            elif port.startswith('lt'):  # lt
                line = re.sub(port, 'lt {}'.format(_translated_ports), line, flags=re.I)
            elif port.startswith('le'):  # le
                line = re.sub(port, 'le {}'.format(_translated_ports), line, flags=re.I)
            elif port.startswith('n'):  # neq
                line = re.sub(port, 'neq {}'.format(_translated_ports), line, flags=re.I)
            elif port.startswith('r'):  # range
                line = re.sub(port, 'range {}'.format(_translated_ports), line, flags=re.I)
            else:
                raise ValueError('Invalid Port: {0} in ACE: {1}'.format(port, self.ace))

        logging.debug('ACE "{0}" translated to "{1}"'.format(self.ace, line))
        return line


cisco_acl/convert_mask.py
import ipaddress
import re
from cisco_acl.regexes import ace_match


def translate_mask(acl_lines, from_type, to_type):
    """ Translate between various mask definitions in ACEs
    
    Args:
        acl_lines: list of ACEs
        from_type: "wc" or "subnet"
        to_type: "wc", "subnet", or "cidr"

    Returns:
        list of ACEs with subnet masks translated
        
    Examples:
        >>> acl_lines = ['permit tcp 10.0.1.0 0.0.0.255 any eq 443']
        >>> translate_mask(acl_lines, 'wc', 'subnet')
        ['permit tcp 10.0.1.0 255.255.255.0 any eq 443']
        >>> translate_mask(acl_lines, 'wc', 'cidr')
        ['permit tcp 10.0.1.0/24 any eq 443']
        >>> acl_lines = ['permit tcp 172.16.1.0 255.255.255.0 any eq 80']
        >>> translate_mask(acl_lines, 'subnet', 'wc')
        ['permit tcp 172.16.1.0 0.0.0.255 any eq 80']
        >>> translate_mask(acl_lines, 'subnet', 'cidr')
        ['permit tcp 172.16.1.0/24 any eq 80']
    """

    output_lines = []
    types = ['wc', 'subnet', 'cidr']
    if from_type not in types or to_type not in types:
        raise TypeError

    # determine if we have a subnet in the ACL line
    subnet_regex = '(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})\s+(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'
    subnets_to_process = {}

    for acl_line in acl_lines:
        output_line = acl_line
        m = ace_match(acl_line.strip())
        if m:
            source_host = m['source']
            destination_host = m['destination']

            for network in [source_host, destination_host]:
                if re.match(subnet_regex, network):
                    try:
                        ip_object = ipaddress.ip_network('/'.join(network.split()))
                        subnets_to_process[network] = ip_object
                    except ValueError as e:
                        continue

        for subnet in subnets_to_process:
            if from_type == 'wc' and to_type == 'cidr':
                output_line = re.sub(subnet, subnets_to_process[subnet].with_prefixlen, output_line)
            elif from_type == 'wc' and to_type == 'subnet':
                output_line = re.sub(subnet, ' '.join(subnets_to_process[subnet].with_netmask.split('/')), output_line)
            elif from_type == 'subnet' and to_type == 'wc':
                output_line = re.sub(subnet, ' '.join(subnets_to_process[subnet].with_hostmask.split('/')), output_line)
            elif from_type == 'subnet' and to_type == 'cidr':
                output_line = re.sub(subnet, subnets_to_process[subnet].with_prefixlen, output_line)

        output_lines.append(output_line)
    return output_lines
